### ISSUE TITLE
Long build times of ensembldb with newer RSQLite packages

### ISSUE THREAD

**Issue by jorainer** [jorainer/ensembldb#11]
Build times differ considerably between `RSQLite` version 1.0.0 and release candidate.
- [ ] Check unit tests.
- [ ] Check examples.
- [ ] Check vignette.

**Comment by jorainer** [comment]
That's no fun.

`R CMD check` of `ensembldb` 1.5.10 with `RSQLite` version 1.0.0:

```
real    6m26.891s
user    5m20.351s
sys 0m35.352s
```

`R CMD check` of `ensembldb` 1.5.10 with `RSQLite` version 1.0.9011:

```
real    12m37.825s
user    9m43.063s
sys 2m25.617s
```

Doesn't seem to be related to a simple `SELECT` query:
`RSQLite` 1.0.0

``` r
> library(EnsDb.Hsapiens.v75)
> edb <- EnsDb.Hsapiens.v75
> system.time(g <- genes(edb))
   user  system elapsed 
  0.440   0.061   0.515 
```

And with `RSQLite` 1.0.9011:

``` r
> library(EnsDb.Hsapiens.v75)
> edb <- EnsDb.Hsapiens.v75
> system.time(g <- genes(edb))
   user  system elapsed 
  0.501   0.048   0.560 
```

**Comment by krlmlr** [comment]
Thanks for looking into it.

**Comment by jorainer** [comment]
Now that's puzzling; I checked the timings of some of the commands for `RSQLite` 1.0.0 and 1.0.9011 and it seems that there is hardly any difference:

`RSQLite` 1.0.0, `ensembldb` 1.5.10:

```
 |                                                        | user.self | sys.self | elapsed |
 |--------------------------------------------------------+-----------+----------+---------|
 | Plain genes call.                                      |     0.545 |    0.064 |   0.625 |
 | genes with GenenameFilter.                             |     0.033 |    0.006 |   0.039 |
 | genes with GenenameFilter and % wildcard.              |     0.029 |    0.005 |   0.034 |
 | genes with SeqnameFilter.                              |     0.047 |    0.005 |   0.051 |
 | genes with SeqnameFilter and GenebiotypeFilter.        |     0.034 |    0.004 |   0.039 |
 | Plain transcripts call.                                |     3.382 |    0.457 |   3.917 |
 | transcripts with a GenenameFilter.                     |     0.252 |    0.025 |   0.279 |
 | transcripts with a GenenameFilter and TxbiotypeFilter. |     0.153 |    0.023 |   0.179 |
 | Plain exons call.                                      |    12.401 |    2.856 |  15.516 |
 | exons with ExonrandFilter.                             |     3.658 |    0.962 |   4.666 |
 | exons with GenenameFilter.                             |     0.171 |    0.026 |   0.203 |
 | Plain transcriptsBy                                    |     2.417 |    0.417 |   2.897 |
 | Plain exonsBy                                          |    20.362 |    4.206 |  25.428 |

```

`RSQLite` 1.0.9011, `ensembldb` 1.5.10:

```
 |                                                        | user.self | sys.self | elapsed |
 |--------------------------------------------------------+-----------+----------+---------|
 | Plain genes call.                                      |     0.538 |    0.054 |   0.609 |
 | genes with GenenameFilter.                             |     0.031 |    0.005 |   0.035 |
 | genes with GenenameFilter and % wildcard.              |     0.037 |    0.006 |   0.043 |
 | genes with SeqnameFilter.                              |     0.046 |    0.005 |   0.051 |
 | genes with SeqnameFilter and GenebiotypeFilter.        |     0.036 |    0.004 |   0.040 |
 | Plain transcripts call.                                |     3.844 |    0.659 |   4.576 |
 | transcripts with a GenenameFilter.                     |     0.216 |    0.025 |   0.242 |
 | transcripts with a GenenameFilter and TxbiotypeFilter. |     0.127 |    0.022 |   0.148 |
 | Plain exons call.                                      |    11.141 |    2.843 |  14.181 |
 | exons with ExonrandFilter.                             |     3.387 |    1.044 |   4.478 |
 | exons with GenenameFilter.                             |     0.029 |    0.005 |   0.034 |
 | Plain transcriptsBy                                    |     2.702 |    0.652 |   3.425 |
 | Plain exonsBy                                          |    17.643 |    3.332 |  21.148 |

```

**Comment by krlmlr** [comment]
I'll double-check on my machine, could be a time measurement problem.

**Comment by jorainer** [comment]
Something very odd is going on; when I render the vignette with 1.0.9011 installed it takes about the same time than with 1.0.0 installed, **BUT** calling `R CMD build ensembldb` takes 20 seconds longer with the newer version.

**Comment by krlmlr** [comment]
Now 470 seconds for R CMD check with current master, and 290 seconds with 1.0.0. Checks now done on an otherwise idle machine.

**Comment by jorainer** [comment]
I'll check the individual unit tests; there most be something there.

**Comment by jorainer** [comment]
OK, identified the culprit (`elapsed_rc` is the timing for the release candidate, `elapsed_sr` for the stable release):

```
 | file_name                  | fun                                | elapsed_rc | elapsed_sr |
 |----------------------------+------------------------------------+------------+------------|
 | test_Functionality.R       | test_UTRs                          |     53.432 |      7.449 |
 | test_Functionality.R       | test_cdsBy                         |     35.823 |      6.262 |
 | test_Filters.R             | test_SymbolFilter                  |      9.968 |      4.980 |
 | test_ucscChromosomeNames.R | test_cdsBy_seqnames                |      5.272 |      4.131 |
 | test_xByOverlap.R          | test_exonsByOverlaps               |     12.917 |     12.163 |
 | test_returnCols.R          | test_with_exons                    |      4.861 |      4.207 |
```

Will now check the individual functions and the code; let's see what's in there.

**Comment by jorainer** [comment]
@krlmlr I got it I think: an SQL call with a `order by` if combined with `case when ... then ... end` condition and a `where` condition is considerably slower in the 1.0.9011 release candidate than in the version 1.0.0. To explain:

``` r
library(EnsDb.Hsapiens.v75)
library(RSQLite)
edb <- EnsDb.Hsapiens.v75
cols <- c("tx_id", "tx_cds_seq_start", "tx_cds_seq_end", "seq_name",
          "seq_strand", "exon_idx", "exon_id", "exon_seq_start",
          "exon_seq_end")
orderBy <- paste0("tx.tx_id, case when seq_strand = 1 then tx_seq_start",
                  " when seq_strand = -1 then (tx_seq_end * -1) end")
the_query <- ensembldb:::.buildQuery(edb, columns = cols,
                                     filter = list(SeqnameFilter("Y")),
                                     order.by = orderBy, skip.order.check = TRUE)

```

Now, this `the_query` joins the tables `gene` with `tx` with `tx2exon` with `exon` which generates a rather huge table. The `where` condition subsets this and the result is then in addition ordered by `tx_id` and `tx_seq_start` when `seq_strand = -1` and by `tx_seq_end * -1` otherwise.

``` r
> the_query
[1] "select distinct gene.seq_name,gene.seq_strand,tx.tx_id,tx.tx_cds_seq_start,tx.tx_cds_seq_end,tx2exon.exon_id,tx2exon.exon_idx,exon.exon_seq_start,exon.exon_seq_end from gene join tx on (gene.gene_id=tx.gene_id) join tx2exon on (tx.tx_id=tx2exon.tx_id) join exon on (tx2exon.exon_id=exon.exon_id) where gene.seq_name = 'Y' order by tx.tx_id, case when seq_strand = 1 then tx_seq_start when seq_strand = -1 then (tx_seq_end * -1) end asc"
```

The timings of this call on `RSQLite` 1.0.0:

``` r
> system.time(res <- dbGetQuery(db_con, the_query))
   user  system elapsed 
  0.725   0.217   0.952 
```

And on `RSQLite` 1.0.9011:

``` r
> system.time(res <- dbGetQuery(db_con, the_query))
   user  system elapsed 
  4.420   1.850   6.303 
```

I could imagine that the query optimizer in the newer RSQLite (or rather SQLite) first orders the results before applying the `where` condition. @krlmlr what do you think?

To fix this I'm thinking of dropping the `order by` in the SQL call in favor of sorting in `R`.

**Comment by krlmlr** [comment]
We did update the sqlite3 library, it might be worthwhile to check the sqlite3 news and/or file an upstream bug report. This could be a regression that has been fixed already.

Can `seq_strand` assume other values? Either way, consider trying `(seq_strand = 1) * tx_seq_start + (seq_strand = -1) * (tx_seq_end * -1)` as a sorting expression.

**Comment by jorainer** [comment]
Thanks for the suggestion, but it didn't improve.

**Comment by jorainer** [comment]
Just an update: I'm going to optimize here and there the code and the queries to speed up things.

**Comment by jorainer** [comment]
With the updated `EnsDb` packages (available in the current BioC 3.4 devel) which are available since today the build times are faster than before.
